frontend/src/components/tabs/MetricsTab.test.tsx

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2021 The Kubeflow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Artifact, ArtifactType, Execution, Value } from '@kubeflow/frontend';
+import { render, waitFor } from '@testing-library/react';
+import { Struct } from 'google-protobuf/google/protobuf/struct_pb';
+import React from 'react';
+import { QueryClient, QueryClientProvider } from 'react-query';
+import { BrowserRouter } from 'react-router-dom';
+import { MetricsTab } from './MetricsTab';
+
+describe('MetricsTab', () => {
+  const mlmdUtils = require('src/lib/MlmdUtils');
+  const OutputArtifactLoader = require('src/lib/OutputArtifactLoader');
+  const execution = new Execution();
+  const queryClient = new QueryClient();
+
+  beforeEach(() => {
+    const executionGetIdSpy = jest.spyOn(execution, 'getId');
+    executionGetIdSpy.mockReturnValue(123);
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
```

Comment:
I think `resetAllMocks` only reset mock state, but the mocks are kept.
However, `restoreAllMocks` remove the mocks entirely back to the original module.

https://jestjs.io/docs/jest-object#jestrestoreallmocks
Therefore, I think we recommend always restoring mocks to avoid issues that cause different test cases interfere with each other. To consistently apply these best practices, suggest directly calling https://github.com/kubeflow/pipelines/blob/226079338b38b85ae3b63a5e289c6ca0cd9511d0/frontend/src/TestUtils.ts#L165 to set up the standard beforeEach calls.

Reply:
Thank you for pointing it out! Using testBestPractices() here.